src/rules/module-folder-structure.ts

```diff
@@ -0,0 +1,140 @@
+
+      if (requiredFolderPatterns.length === 0) {
+        continue;
+      }
+
+      // build the required file paths relative to module path
+      const { path } = module;
+
+      const fgConfig: Options = {
+        cwd: path,
+        onlyDirectories: true,
+        globstar: true,
+        extglob: true,
+        dot: true,
+      };
+
+      const allModuleFolders = fg.sync('**/*', fgConfig);
+
+      // Evaluate the pattern one by one and check if there is any missing a match
+      for (const requiredFolderPattern of requiredFolderPatterns) {
+        const foldersMatching = fg.sync(requiredFolderPattern, fgConfig);
+
+        if (foldersMatching.length) {
+          results.push({
+            valid: true,
+            resource: requiredFolderPattern,
+            message: 'Required folder found',
+            rule: rule.name,
+            module,
+          });
+        } else {
+            // If there is no match, then it's missing a required folder for this pattern
+          results.push({
+            valid: false,
+            resource: requiredFolderPattern,
+            message: 'Required folder not found',
+            rule: rule.name,
+            module,
+          });
+        }
+      }
+
+      if (strict) {
+        const foldersMatchingAllPatterns = fg.sync(requiredFolderPatterns, fgConfig);
+
+        // ? If there's no match for all patterns, then we shouldn't check extra folders
+        if (foldersMatchingAllPatterns.length === 0) {
+          continue;
+        }
+
+        const extraFolders = allModuleFolders.filter((moduleFolder) => (
+          !foldersMatchingAllPatterns.some((matchedFolder) => matchedFolder === moduleFolder || matchedFolder.startsWith(`${moduleFolder}/`))
+        ));
+
+        if (extraFolders.length) {
+          extraFolders.forEach((folder) => {
+            results.push({
+              valid: false,
+              resource: folder,
+              message: 'Folder outside the required list not allowed (strict mode)',
+              rule: rule.name,
+              module,
+            });
+          });
+        }
+      }
+    }
+
+    return results;
+  },
+  check() {
+    return null;
+  },
+  docMarkdown() {
+    return '* Check whether the folder structure is present in the modules';
+  },
+  docExampleConfigs() {
+    return [
+      {
+        description: 'Deactivates this rule',
+        config: false,
+      },
+      {
+        description: 'Activates this rule using default folders (defaults: `["src"]`)',
+        config: true,
+      },
+      {
+        description: 'Loosely requires module structure. The module should contain **at least** this set of folders, but can still have others.',
+        config: {
+          strict: false,
+          folders: ['src', 'docs', 'libs'],
+        },
+      },
+      {
+        description: 'Strictly requires module structure. No extra folders allowed, should match exactly.',
```

Comment:
Writing style discussion: the text is very well written, but maybe you could describe it more as a concrete "example" of what it's doing ("documentation by examples"), like:

"Allow only folder 'src/test' or any folders inside 'src' that has a name 'util'".

We've been writing like this in other rules. Maybe a combination of both writing styles can help people understand. What do you think?

Reply:
I agree! I was mainly inspired by the writing style from the `required-files` rule. Maybe we could revisit them all at a later stage.